src/librustc/middle/cstore.rs

```diff
@@ -474,7 +477,13 @@ pub mod tls {
     {
         let tls_payload = (ecx as *const _, encoder as *mut _);
         let tls_ptr = &tls_payload as *const _ as *const TlsPayload;
-        TLS_ENCODING.set(unsafe { &*tls_ptr }, || f(ecx, encoder))
+        TLS_ENCODING.with(|tls| {
+            let prev = tls.get();
```

Comment:
Ok, if anyone's surprised about these impls, I'd be more than happy to submit a PR with a scoped-tls-like abstraction to one of the rustc libs we have! For now I'll try to land this ahead of time.